index.js

```diff
@@ -113,66 +113,103 @@ class CarmiObserver extends React.Component {
+  const {type, props, childrenList, extraProps, key} = parseDescriptor(descriptor);
+  const privates = getPrivatesByInstance(this);
+  const currentElement = privates.descriptorToElementsMap.get(descriptor);
+  if (currentElement && currentElement.props.type === type && getMaybeKey(currentElement.props, 'origKey') === key) {
+    // Element is mounted
     if (privates.root && privates.descriptorToCompsMap.has(descriptor)) {
       privates.descriptorToCompsMap.get(descriptor).forEach(comp => privates.pendingFlush.add(comp));
     } else {
-      prevElement.props.dirtyFlag[0] = true;
+      currentElement.props.dirtyFlag[0] = true;
     }
-    Object.assign(prevElement.props, childExtraProps);
-    Object.keys(prevElement.props).forEach(prop => {
-      if (!childExtraProps.hasOwnProperty(prop) && !BUILT_IN_PROPS.hasOwnProperty(prop)) {
-        delete prevElement.props[prop];
-      }
-    });
+    replaceElementProps(currentElement, extraProps);
   } else {
     const dirtyFlag = [true];
-    const props = { descriptor, type, dirtyFlag };
+    const observerProps = {descriptor, type, dirtyFlag, token: pointerByInstance.get(this)};
     if (key !== null) {
-      props.origKey = key;
-      props.key = key;
+      observerProps.origKey = key;
+      observerProps.key = key;
     }
-    const rawElement = React.createElement(React.forwardRef((forwardProps, forwardedRef) => {
-      let overrides = null;
-      Object.keys(forwardProps).forEach(prop => {
-        if (!BUILT_IN_PROPS.hasOwnProperty(prop) && forwardProps[prop] !== childProps[prop]) {
-          overrides = overrides || {};
-          overrides[prop] = forwardProps[prop];
-        }
-      })
-      return React.createElement( CarmiObserver, {...props,overrides, forwardedRef})
-    }), { ...props, ...childExtraProps });
-    // sorry about doing it but 
-    // we short circuit the reconcilation code of React
-    // and we need to mutate the props in place
-    // so a parent component that is aware of the expected props
-    // of a child can access them, even if they changed since the
-    // last time React.createElement was triggered
-    const element = { ...rawElement }
-    element.props = { ...rawElement.props };
-    privates.descriptorToElementsMap.set(descriptor, element);
+    const newElement = createNewElement(observerProps, props);
+    privates.descriptorToElementsMap.set(descriptor, newElement);
   }
-  return privates.descriptorToElementsMap.get(descriptor);
+  const element = privates.descriptorToElementsMap.get(descriptor);
+  return wrapElement(wrappers, element, childrenList);
+}
+
+function getOverrides(originalProps, props) {
+  let overrides = null;
+  Object.keys(props).forEach(prop => {
+    if (!BUILT_IN_PROPS.hasOwnProperty(prop) && originalProps[prop] !== props[prop]) {
+      overrides = overrides || {};
+      overrides[prop] = props[prop];
+    }
+  })
+  return overrides;
+}
+
+function createNewElement(observerProps, props) {
```

Comment:
Might be nit picking, but it could be nicer to have this method showing before the createElement to avoid relying on hoisting.